File: PSET1/greedy.c
#include <stdio.h>
#include <cs50.h>

#define QUARTER 25
#define DIME 10
#define NICKEL 5
#define PENNY 1
#define CORRECT_IMPRECISION 0.0001

int main(void)
{
    float change_input;

    do {
        printf("O hai! How much change is owed?\n");
        change_input = get_float();
    }
    while (change_input < 0);

    int change_owned = (int) ((change_input + CORRECT_IMPRECISION) * 100); // Normalise floats (e.g '0.409995' ==> '41')

    int total_coins = 0;

    while (change_owned > 0) {

        if (change_owned - QUARTER >= 0) {
            change_owned -= QUARTER;
        }
        else if (change_owned - DIME >= 0) {
            change_owned -= DIME;
        }
        else if (change_owned - NICKEL >= 0) {
            change_owned -= NICKEL;
        }
        else {
            change_owned -= PENNY;
        }
        
        total_coins++;
    }

    printf("%d\n", total_coins);

    return 0;
}

File: PSET6/greedy.py
#!/usr/bin/env python3


def main():
    change_owed = get_change()

    min_coins = get_min_coins(change_owed)

    print(min_coins)


def get_change():
    print("O hai! ", end="")
    while True:
        try:
            change = float(input("How much change is owed?\n"))
        except ValueError:
            continue
        if change > 0:
            break

    return change


def get_min_coins(change):
    min_coins = 0
    change *= 100

    quarter = 25
    dime = 10
    nickel = 5
    penny = 1

    while change > 0:
        if change - quarter >= 0:
            change -= quarter
            min_coins += 1
        elif change - dime >= 0:
            change -= dime
            min_coins += 1
        elif change - nickel >= 0:
            change -= nickel
            min_coins += 1
        else:
            change -= penny
            min_coins += 1

    return min_coins


if __name__ == "__main__":
    main()
